ARC56: Extended App Description (#258)
algorandfoundation/ARCs: ARCs/arc-0056.md

  /** All of the methods that the contract implements */
  methods: Method[];
  state: {
    /** Defines the values that should be used for GlobalNumUint, GlobalNumByteSlice, LocalNumUint, and LocalNumByteSlice when creating the application  */
    schema: {
      global: {
        ints: number;
        bytes: number;
      };
      local: {
        ints: number;
        bytes: number;
      };
    };
    /** Describes single key-value pairs in the application's state */
    keys: {
      global: StorageKey[];
      local: StorageKey[];
      box: StorageKey[];
    };
    /** Describes key-value maps in the application's state */
    maps: {
      global: StorageMap[];
      local: StorageMap[];
      box: StorageMap[];
    };
  };
  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
  bareActions: {
    /** OnCompeltes this method allows when appID === 0 */
    create: ('NoOp' | 'OptIn' | 'DeleteApplication')[];
    /** OnCompeltes this method allows when appID !== 0 */
    call: ('NoOp' | 'OptIn' | 'CloseOut' | 'ClearState' | 'UpdateApplication' | 'DeleteApplication')[];
  };

>   /** A mapping of program counters to human-readable error strings */

Reviewer: ```suggestion
  /** A mapping of program counters to human-readable error strings. Error strings SHOULD match program counter's comments in TEAL source.*/
```

---
arc: 56
title: Extended App Description
description: Adds information to the ABI JSON description
author: Joe Polny (@joe-p)
discussions-to: https://github.com/algorandfoundation/ARCs/issues/258
status: Draft
type: Standards Track
category: ARC
created: 2023-11-14

> requires: 4

Reviewer: ```suggestion
requires: 4, 28
```

type: Standards Track
category: ARC
created: 2023-11-14
requires: 4, 22, 28
---

## Abstract
This ARC takes the existing JSON description of a contract as described in [ARC-4](./arc-0004.md) and adds more fields for the purpose of client interaction

## Motivation
The data provided by ARC-4 is missing a lot of critical information that clients should know when interacting with an app. This means ARC-4 is insufficient to generate type-safe clients that provide a superior developer experience.

On the other hand, [ARC-32](./arc-0032.md) provides the vast majority of useful information that can be used to <a href="https://github.com/algorandfoundation/algokit-cli/blob/main/docs/features/generate.md#1-typed-clients">generate typed clients</a>, but requires a separate JSON file on top of the ARC-4 json file, which adds extra complexity and cognitive overhead.

## Specification

### Contract Interface
Every application is described via the following interface which is an extension of the `Contract` interface described in [ARC-4](./arc-0004.md).

```ts
/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
interface Contract {

>   /** The ARCs used and/or supported by this contract */

Reviewer 2: Worth indicating the default should be `[4, 28, 56]`?

  /** All of the methods that the contract implements */
  methods: Method[];
  state: {
    /** Defines the values that should be used for GlobalNumUint, GlobalNumByteSlice, LocalNumUint, and LocalNumByteSlice when creating the application  */
    schema: {
      global: {
        ints: number;
        bytes: number;
      };
      local: {
        ints: number;
        bytes: number;
      };
    };
    /** Describes single key-value pairs in the application's state */
    keys: {
      global: StorageKey[];
      local: StorageKey[];
      box: StorageKey[];
    };
    /** Describes key-value maps in the application's state */
    maps: {
      global: StorageMap[];
      local: StorageMap[];
      box: StorageMap[];
    };
  };
  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
  bareActions: {
    /** OnCompletes this method allows when appID === 0 */
    create: ('NoOp' | 'OptIn' | 'DeleteApplication')[];
    /** OnCompletes this method allows when appID !== 0 */
    call: ('NoOp' | 'OptIn' | 'CloseOut' | 'ClearState' | 'UpdateApplication' | 'DeleteApplication')[];
  };
  /** Information about the TEAL */

>   sourceInfo: SourceInfo;

Reviewer 2: I think this should be optional so you can distribute a spec file to allow users to talk to a deployed contract without needing to have the TEAL.

>   sourceInfo: SourceInfo;

Reviewer 2: This should be optional?

Reviewer 2: Also, should it be an array?

    struct?: StructName;
    /** Optional, user-friendly description for the return value */
    desc?: string;
  };
  /** an action is a combination of call/create and an OnComplete */
  actions: {
    /** OnCompletes this method allows when appID === 0 */
    create: ('NoOp' | 'OptIn' | 'DeleteApplication')[];
    /** OnCompletes this method allows when appID !== 0 */
    call: ('NoOp' | 'OptIn' | 'CloseOut' | 'ClearState' | 'UpdateApplication' | 'DeleteApplication')[];
  };
  /** If this method does not write anything to the ledger (ARC-22) */
  readonly: boolean;
  /** ARC-28 events that MAY be emitted by this method */
  events?: Array<Event>;
  /** Information that clients can use when calling the method */
  recommendations?: {
    /** The number of inner transactions the caller should cover the fees for */
    innerTransactionCount?: number;
    /** Recommended box references to include */
    boxes?: {
      /** The app ID for the box */
      app: number;
      /** The base64 encoded box key */
      key: string;
      /** The number of bytes being read from the box */
      readBytes: number;
      /** The number of bytes being written to the box */

>       writeBytes: number;

Reviewer 2: should readBytes and writeBytes and app be optional?

    schema: {
      global: {
        ints: number;
        bytes: number;
      };
      local: {
        ints: number;
        bytes: number;
      };
    };
    /** Mapping of human-readable names to StorageKey objects */
    keys: {
      global: { [name: string]: StorageKey };
      local: { [name: string]: StorageKey };
      box: { [name: string]: StorageKey };
    };
    /** Mapping of human-readable names to StorageMap objects */
    maps: {
      global: { [name: string]: StorageMap };
      local: { [name: string]: StorageMap };
      box: { [name: string]: StorageMap };
    };
  };
  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
  bareActions: {
    /** OnCompletes this method allows when appID === 0 */
    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
    /** OnCompletes this method allows when appID !== 0 */
    call: (
      | "NoOp"
      | "OptIn"
      | "CloseOut"
      | "ClearState"
      | "UpdateApplication"
      | "DeleteApplication"
    )[];
  };
  /** Information about the TEAL */

>   sourceInfo?: SourceInfo[];

Reviewer 3: Where is `SourceInfo` defined?

  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23, but otherwise MUST be defined. */
  source?: {
    /** The approval program */

>     approval: string;

Reviewer 3: can we make this be the TEAL or the bytecode?

Author: Are you saying have two properties? One for pre-compiled TEAL and one for compiled bytecode?

    /** The clear program */
    clear: string;
  };
  /** ARC-28 events that MAY be emitted by this contract */
  events?: Array<Event>;

>   /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) and their respecive types */

Reviewer 3: There doesn't seem to be any harm if we include the `TMPL_` prefix? The less assumptions (within the bounds of practicality) the better, no?

Reviewer 2: But if we use it for code generation then you’d have to strip the TMPL so we have to cater for it somewhere?

Reviewer 3: If certain tooling expects that as a prefix, that seems like the place to check for and strip that prefix, yeah?

Reviewer 2: *shrug* could go either way,  I guess it's fair that this could be more general purpose and then the generators are clever enough to detect and strip TMPL_ when emitting code...

Author: My rationale omitting the TMPL_ here and forcing tooling to use it is to make it easier to implement safely. For example, if ARC56 allowed anything as the variable name and the tooling didn't have to add `TMPL_`, it would need to check all the variable names against all opcodes. Otherwise an ARC56 file could have a template variable named `byte` which subsequently overwrites all `byte` opcodes which is not the intended effect of template variables.

  templateVariables?: {
    [name: string]: ABIType | AVMBytes | StructName;
  };
}
```

### Method Interface

Every method in the contract is described via a `Method` interface. This interface is an extension of the one defined in [ARC-4](./arc-0004.md).

```ts
/** Describes a method in the contract. This interface is an extension of the interface described in ARC-4 */
interface Method {
  /** The name of the method */
  name: string;
  /** Optional, user-friendly description for the method */
  desc?: string;
  /** The arguments of the method, in order */
  args: Array<{
    /** The type of the argument */

>     type: ABIType;

Reviewer 3: What should the value of `type` be if `struct` is also defined?

Reviewer 3: Ah, the equivalent ARC-4 tuple type string I guess?

Author: Yes just the ABI type for backwards compat with ARC4. ARC56 parsers read `struct ?? type` to determine the ARC56 type to use

Reviewer 2: ah, maybe this is the answer to the question I just asked. It wasn't immediately obvious to me that I should do that. We should probably indicate it in the description comments at the very least to make it more obvious for implementers?

Reviewer 2: Same detail for returns...

  /** Information that clients can use when calling the method */
  recommendations?: {
    /** The number of inner transactions the caller should cover the fees for */
    innerTransactionCount?: number;
    /** Recommended box references to include */
    boxes?: {
      /** The app ID for the box */
      app?: number;
      /** The base64 encoded box key */
      key: string;
      /** The number of bytes being read from the box */
      readBytes: number;
      /** The number of bytes being written to the box */
      writeBytes: number;
    };
    /** Recommended foreign accounts */
    accounts?: string[];
    /** Recommended foreign apps */
    apps?: number[];
    /** Recommended foreign assets */
    assets?: number[];
  };
}
```

### Event Interface

[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments

```ts
interface Event {
  /** The name of the event */
  name: string;
  /** Optional, user-friendly description for the event */
  desc?: string;
  /** The arguments of the event, in order */
  args: Array<{
    /** The type of the argument */

>     type: ABIType;

Reviewer 3: Since this isn't part of ARC-4, maybe in the case of a struct, we can avoid re-specifying it as a tuple here?

Author: It is not a part of ARC4 but it is a part of ARC28. You could make the argument that ARC28 doesn't have much adoption yet, but I think it's better to keep backwards compatibility with existing ARCs

    /** Optional, user-friendly name for the argument */
    name?: string;
    /** Optional, user-friendly description for the argument */
    desc?: string;
    /** If the type is a struct, the name of the struct */
    struct?: StructName;
  }>;
}
```

### Type Interfaces

The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.

```ts
/** An ABI-encoded type */
type ABIType = string;

/** The name of a defined struct */
type StructName = string;

/** Raw byteslice without the length prefixed that is specified in ARC-4 */
type AVMBytes = "bytes";

/** Mapping of named structs to the ABI type of their fields */
interface StructFields {
  [fieldName: string]: ABIType | StructFields;
}
```

### Storage Interfaces

These interfaces properly describe how app storage is access within the contract

```ts
/** Describes a single key in app storage */
interface StorageKey {
  /** Description of what this storage key holds */

>   desc?: string;

Reviewer 3: Can we use `description` here and elsewhere? I know beaker used `descr` and so did ARC-32 naturally, but other things have used `desc` like here, and it can get confusing. Pretty sure I recall there being at least one bug in a client lib relating to this?

Reviewer 2: Haha I raised the same thing :p 
The thinking was that arc-4 has desc so this is consistent. The alternative would be breaking changes to arc-4 which felt unpalatable

Reviewer 3: Ah, I should read to the end first - it's for backwards compatibility with ARC-4, got it.

  /** The type of the key */

>   keyType: ABIType | AVMBytes | StructName;

Reviewer 3: Why does StorageKey have keyType? Isn't this for singular, known values?

Author: Yes but I figured it's still useful to know the type for things like explorers and frontends

Reviewer 3: So it's intended to indicate the encoding of `key`? Maybe `keyEncoding` would be clearer? I don't think it benefits from consistency of naming with `StorageMap.keyType` since that seems of a different purpose - to indicate what sort of value to encode to combine with `StorageMap.key` in order to retrieve a value. Whereas `StorageKey.key` is already encoded and you may want to know how to interpret that value. So they're kind of opposites?

Reviewer 2: Yeah, I'm a bit confused about what `keyType` means?

Reviewer 2: Oh wait, I think I see.

So what should this and the `key` value be for a string value like we had in arc-32?

Needing to have key as a base64 reduces the human readability of these files... I wonder if keyType should allow for a raw string to be used?

Reviewer 2: I think this would be the equivalent?

```typescript
          key: Buffer.from('my key', 'utf-8').toString("base64"),
          keyType: 'bytes',
```

Reviewer 2: But then that doesn't let explorers etc. know that the key is actually a (non-ARC4) string?

Author: I think it would be confusing to have some keys be ASCII encoded and some keys be base64, thus everything base64 makes the most sense to me. I'm not sure if human readability is all that important. If you want to make it human readable just through it into Lora (once supported).

Reviewer 2: Human readable is a nice feature though - makes it easy to see things at a glance, I've certainly found that useful in the past.

Regardless if that's supported or not, it probably makes sense to support a keyType of `utf-8`, which would either mean non-base64 or base64 `bytes`, but representing a UTF-8 string so the likes of Lora can show it properly.

Also, similarly we should probably have a key type of avm_address or similar to denote when you use the public key bytes of an address directly (i.e. non-ARC4 encoding).

  /** The type of the value */
  valueType: ABIType | AVMBytes | StructName;
  /** The bytes of the key encoded as base64 */
  key: string;
}

/** Describes a mapping of key-value pairs in storage */
interface StorageMap {
  /** Description of what the key-value pairs in this mapping hold */
  desc?: string;
  /** The type of the keys in the map */
  keyType: ABIType | AVMBytes | StructName;
  /** The type of the values in the map */
  valueType: ABIType | AVMBytes | StructName;
  /** The prefix of the map, encoded as a utf-8 string */

>   prefix?: string;

Reviewer 3: Why `utf-8` here and not `base64` like `StorageKey.key`? UTF-8 is more readable, but does that not preclude raw byte prefixes?

Author: Good point. I can change to base64

type: Standards Track
category: ARC
created: 2023-11-14
requires: 4, 22, 28
---

## Abstract

This ARC takes the existing JSON description of a contract as described in [ARC-4](./arc-0004.md) and adds more fields for the purpose of client interaction

## Motivation

The data provided by ARC-4 is missing a lot of critical information that clients should know when interacting with an app. This means ARC-4 is insufficient to generate type-safe clients that provide a superior developer experience.

On the other hand, [ARC-32](./arc-0032.md) provides the vast majority of useful information that can be used to <a href="https://github.com/algorandfoundation/algokit-cli/blob/main/docs/features/generate.md#1-typed-clients">generate typed clients</a>, but requires a separate JSON file on top of the ARC-4 json file, which adds extra complexity and cognitive overhead.

## Specification

### Contract Interface

Every application is described via the following interface which is an extension of the `Contract` interface described in [ARC-4](./arc-0004.md).

```ts
/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
interface Contract {

>   /** The ARCs used and/or supported by this contract. All contracts implicity support ARC4 and ARC56 */

Reviewer 2: ```suggestion
  /** The ARCs used and/or supported by this contract. All contracts implicity support ARC-4, ARC-28 and ARC-56 */
```

Author: I didn't put ARC-28 and ARC-22 here by default because a contract might not use events or readonly at all, in which case it would seem weird to say every contract supports them. I also don't think it's technically wrong, but if we want to include ARC-28 here we should also include ARC-22.

Reviewer 2: That's fair. Need to keep it in mind for Puya to emit those ARCs if those features are used then. cc @achidlow

  arcs: number[];
  /** A user-friendly name for the contract */
  name: string;
  /** Optional, user-friendly description for the interface */
  desc?: string;
  /**
   * Optional object listing the contract instances across different networks
   */
  networks?: {
    /**

>      * The key is the base64 genesis hash of the network, and the value contains

Reviewer 2: This description should probably be added to the networks key above otherwise you don't see it in intellisense

     * information about the deployed contract in the network indicated by the
     * key. A key containing the human-readable name of the network MAY be
     * included, but the corresponding genesis hash key MUST also be defined
     */
    [network: string]: {
      /** The app ID of the deployed contract in this network */

>       appID: number;

Reviewer 2: Everywhere else we use a `appId` convention, should we be consistent?

Author: `appID` is how its defined in ARC4

Using TS interfaces is just a convenient way to define a JSON interface, so I don't think we necessarily need to specify bigint because it doesn't apply to every language.

Reviewer 2: Fair, happy with that then

    schema: {
      global: {
        ints: number;
        bytes: number;
      };
      local: {
        ints: number;
        bytes: number;
      };
    };
    /** Mapping of human-readable names to StorageKey objects */
    keys: {
      global: { [name: string]: StorageKey };
      local: { [name: string]: StorageKey };
      box: { [name: string]: StorageKey };
    };
    /** Mapping of human-readable names to StorageMap objects */
    maps: {
      global: { [name: string]: StorageMap };
      local: { [name: string]: StorageMap };
      box: { [name: string]: StorageMap };
    };
  };
  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
  bareActions: {
    /** OnCompletes this method allows when appID === 0 */
    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
    /** OnCompletes this method allows when appID !== 0 */
    call: (
      | "NoOp"
      | "OptIn"
      | "CloseOut"
      | "ClearState"
      | "UpdateApplication"
      | "DeleteApplication"
    )[];
  };
  /** Information about the TEAL */

>   sourceInfo?: SourceInfo[];

Reviewer 2: Should this be keyed by approval vs clear?

Author: Yeah good call. Will update.

    call: (
      | "NoOp"
      | "OptIn"
      | "CloseOut"
      | "ClearState"
      | "UpdateApplication"
      | "DeleteApplication"
    )[];
  };
  /** Information about the TEAL programs */
  sourceInfo?: {
    /** Approval program information */
    approval: SourceInfo[];
    /** Clear program information */
    clear: SourceInfo[];
  };
  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23 */
  source?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** The compiled bytecode for the application. MUST be omitted if included as part of ARC23 */
  byteCode?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** Information used to get the given byteCode and/or PC values in sourceInfo. MUST be given if byteCode or PC values are present */
  compilerInfo?: {
    /** The name of the compiler */
    compiler: "algod" | "puya";
    /** Compiler version information */
    compilerVersion: {
      major: number;
      minor: number;
      patch: number;

>       commit?: string;

Reviewer 2: tealscript emits `commitHash` rather than `commit`, which I think is a better property name. Should we change?

Author: Yeah I will update the ARC to use `commitHash`

    };
  };
  /** ARC-28 events that MAY be emitted by this contract */
  events?: Array<Event>;
  /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) to their respecive types and values (if applicable) */

>   templateVariables?: {

Reviewer 4: Would it make sense to also allow these templateVariables as part of the networks section (along with app id)?

As part of building puya source maps we realised we need to know the template variables used for a particular contract as this can impact the PC locations, and one of the ideas we had was using ARC-56 as a source of these values.

Reviewer 2: @daniel-makerx how important is this in the end with the new way you were thinking about things? Still useful?

Reviewer 4: Not required for source maps if template values are allowed in constant blocks. 

But may still be potentially useful in some use cases to know by network the values used?
Happy to leave it out tho, could always be added with a subsequent ARC

Reviewer 2: Agreed, let's leave it out for now

    [name: string]: {
      /** The type of the template variable */
      type: ABIType | AVMBytes | StructName;
      /** If given, the the base64 encoded value used for the given app/program */
      value?: string;
    };
  };
  /** The scratch variables used during runtime */
  scratchVariables?: {
    [name: string]: {
      slot: number;
      type: ABIType | AVMBytes | StructName;
    };
  };
}
```

### Method Interface

Every method in the contract is described via a `Method` interface. This interface is an extension of the one defined in [ARC-4](./arc-0004.md).

```ts
/** Describes a method in the contract. This interface is an extension of the interface described in ARC-4 */
interface Method {
  /** The name of the method */
  name: string;
  /** Optional, user-friendly description for the method */
  desc?: string;
  /** The arguments of the method, in order */
  args: Array<{
    /** The type of the argument */
    type: ABIType;
    /** If the type is a struct, the name of the struct */

>     struct?: StructName;

Reviewer 2: Something I just noticed that confused me was that here we have `struct`, but for the `keyType`/`valueType` in storage we don't have that separation, we just have the struct name. That's quite confusing. Should we include `keyStruct` and `valueStruct` to avoid this confusion?

Author: The reason we do this is because the `type` field needs to be backwards compatible with ARC4, but `keyType` and `valueType` don't.

Reviewer 2: Yeah I eventually figured that out. I think we should change the descriptions of all of these fields to be explicit about it all so if you consume arc-56 it’s more obvious. The fact you sometimes use struct and sometimes don’t is a bit confusing.

```ts
interface Event {
  /** The name of the event */
  name: string;
  /** Optional, user-friendly description for the event */
  desc?: string;
  /** The arguments of the event, in order */
  args: Array<{
    /** The type of the argument */
    type: ABIType;
    /** Optional, user-friendly name for the argument */
    name?: string;
    /** Optional, user-friendly description for the argument */
    desc?: string;
    /** If the type is a struct, the name of the struct */
    struct?: StructName;
  }>;
}
```

### Type Interfaces

The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.

```ts
/** An ABI-encoded type */
type ABIType = string;

/** The name of a defined struct */
type StructName = string;

/** Raw byteslice without the length prefixed that is specified in ARC-4 */
type AVMBytes = "bytes";

/** Mapping of named structs to the ABI type of their fields */
interface StructFields {

>   [fieldName: string]: ABIType | StructFields;

Reviewer 2: It occurs to me that this format requires property order to be maintained to be able to map them to the underlying tuple type. Are we comfortable that will always be the case across all languages?

https://www.rfc-editor.org/rfc/rfc7159 says (emphasis mine):

>  An object is an **unordered** collection of zero or more name/value pairs, where a name is a string and a value is a string, number, boolean, null, object, or array.

Author: I think this a choice for the HLL. The struct have to have a specific order for ABI encoding, so the HLL can either implement an ordered struct or not expose ways of iterating keys/values (like in TEALScript)

Reviewer 2: But this spec determines the format of the json file? If the json file is an object and the programming language parsing it doesn’t implemented ordered iteration of object keys you are hosed?

Author: Ah right ok I see what you are saying. So I think it should be `StructFields = { fieldName: string, type: ABIType | StructFields }[]` to preserve order as an array, right?

Reviewer 2: Spot on I think

}
```

### Storage Interfaces

These interfaces properly describe how app storage is access within the contract

```ts
/** Describes a single key in app storage */
interface StorageKey {
  /** Description of what this storage key holds */
  desc?: string;
  /** The type of the key */
  keyType: ABIType | AVMBytes | StructName;
  /** The type of the value */
  valueType: ABIType | AVMBytes | StructName;
  /** The bytes of the key encoded as base64 */
  key: string;
}

/** Describes a mapping of key-value pairs in storage */
interface StorageMap {
  /** Description of what the key-value pairs in this mapping hold */
  desc?: string;
  /** The type of the keys in the map */
  keyType: ABIType | AVMBytes | StructName;
  /** The type of the values in the map */
  valueType: ABIType | AVMBytes | StructName;
  /** The base64-encoded prefix of the map keys*/
  prefix?: string;
}
```

### SourceInfo Interface

This interface gives clients more information about the contract's source code.

```ts

> interface SourceInfo {

Reviewer 2: tealscript also seems to have a `source` property

@daniel-makerx is currently looking at how to represent source maps from high level language, probably worth both of you having a chat and seeing what makes sense?

Reviewer 4: Looks like the `source` property points to the source line in a high level language. It would only make sense to include that in ARC-56 if the source path and contents was included as well. But even that can get tricky as there can be multiple files involved.

If the primary goal of SourceInfo is to provide some basic error/debugging info for TEAL source, then I think what is in the spec currently is sufficient.

With regards to the source maps we are looking at for Puya, they will include more information than just PC -> line mappings, ideally enough so that resolving variables and their types to AVM stack values is possible, as well as preserving call stack information.

Reviewer 2: Agreed.

Reviewer 2: @joe-p thoughts on what we do to resolve this? The typed client TypeScript output doesn't like the extra source property, if we want to include it we should put it in the spec here as an optional?

Author: Ah right yeah... it is really only useful for the developer which will have the full source code available to them. I think adding `source?: string` makes the most sense, with the values the file and line no ie. `somefile.py:1337`

Reviewer 2: I'd also say that that info is mainly useful for breakpoint debugging of HLL source? In which case the work that Daniel and Al are doing will set the "how" (and it's not via ARC-56). Loading that level of debugging info doesn't feel that useful and adds significant weight to the file, should we leave it out?

Author: Breakpoint debugging is useful when you are running tests, but there might be unexpected errors than are encountered when integrating into a frontend or in production in which case you'll need to work backwards from the PC to high-level source

Reviewer 2: Exactly my thinking. Otherwise the arc-56 file is huge. I mean it is anyway with the teal source info, but even more so with HLL source...

Author: Taking it a step further... perhaps for ARC56 it makes sense to only include SoureInfo objects that have an error message

Author: And also omit TEAL line mapping

Reviewer 2: Assuming we are happy with just doing program counter with the implication that means an extra HTTP request sometimes, then yep I think that makes sense.

Author: When would you need to do the extra HTTP request?

Reviewer 2: To grab the deployed byte code (although on reflection, we can stash that away like we do for source maps today so on reflection it's all good actually)

Author: To be clear I don't mean removing source entirely, just the TEAL lines from the mapping. So in the end you just have a mapping of PCs to error messages. 19c62692572f111edd93cf673863ac64bd33c84e

Reviewer 2: Yep understood and I concur :)

Every application is described via the following interface which is an extension of the `Contract` interface described in [ARC-4](./arc-0004.md).

```ts
/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
interface Contract {
  /** The ARCs used and/or supported by this contract. All contracts implicitly support ARC4 and ARC56 */
  arcs: number[];
  /** A user-friendly name for the contract */
  name: string;
  /** Optional, user-friendly description for the interface */
  desc?: string;
  /**
   * Optional object listing the contract instances across different networks.
   * The key is the base64 genesis hash of the network, and the value contains
   * information about the deployed contract in the network indicated by the
   * key. A key containing the human-readable name of the network MAY be
   * included, but the corresponding genesis hash key MUST also be define
   */
  networks?: {
    [network: string]: {
      /** The app ID of the deployed contract in this network */
      appID: number;
    };
  };
  /** Named structs use by the application. Each struct field appears in the same order as ABI encoding. */

>   structs: { [structName: StructName]: StructField[] };

Reviewer 4: Minor thing, noticed both `StructField` and this mapping don't have any sort of `desc` property, might be useful to have?

type: Standards Track
category: ARC
created: 2023-11-14
requires: 4, 22, 28
---

## Abstract

This ARC takes the existing JSON description of a contract as described in [ARC-4](./arc-0004.md) and adds more fields for the purpose of client interaction

## Motivation

The data provided by ARC-4 is missing a lot of critical information that clients should know when interacting with an app. This means ARC-4 is insufficient to generate type-safe clients that provide a superior developer experience.

On the other hand, [ARC-32](./arc-0032.md) provides the vast majority of useful information that can be used to <a href="https://github.com/algorandfoundation/algokit-cli/blob/main/docs/features/generate.md#1-typed-clients">generate typed clients</a>, but requires a separate JSON file on top of the ARC-4 json file, which adds extra complexity and cognitive overhead.

## Specification

### Contract Interface

Every application is described via the following interface which is an extension of the `Contract` interface described in [ARC-4](./arc-0004.md).

```ts
/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
interface Contract {

>   /** The ARCs used and/or supported by this contract. All contracts implicity support ARC4 and ARC56 */

Reviewer 4: ```suggestion
  /** The ARCs used and/or supported by this contract. All contracts implicitly support ARC4 and ARC56 */
```

      | "ClearState"
      | "UpdateApplication"
      | "DeleteApplication"
    )[];
  };
  /** Information about the TEAL programs */
  sourceInfo?: {
    /** Approval program information */
    approval: ProgramSourceInfo;
    /** Clear program information */
    clear: ProgramSourceInfo;
  };
  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23 */
  source?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** The compiled bytecode for the application. MUST be omitted if included as part of ARC23 */
  byteCode?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** Information used to get the given byteCode and/or PC values in sourceInfo. MUST be given if byteCode or PC values are present */
  compilerInfo?: {
    /** The name of the compiler */
    compiler: "algod" | "puya";
    /** Compiler version information */
    compilerVersion: {
      major: number;
      minor: number;
      patch: number;
      commitHash?: string;
    };
  };
  /** ARC-28 events that MAY be emitted by this contract */

>   events?: Array<Event>;

Reviewer 4: Is this to cover any events not already covered by methods? e.g. events emitted by bare actions?

Reviewer 4: Ah I see, I should have re-read ARC-28

>   /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) to their respecive types and values (if applicable) */

Reviewer 4: ```suggestion
  /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) to their respective types and values (if applicable) */
```

  templateVariables?: {
    [name: string]: {
      /** The type of the template variable */
      type: ABIType | AVMType | StructName;
      /** If given, the the base64 encoded value used for the given app/program */
      value?: string;
    };
  };
  /** The scratch variables used during runtime */
  scratchVariables?: {

>     [name: string]: {

Reviewer 4: I assume name is meant to be a variable name?

Wondering if this should be an array with an optional name (and maybe description?) as scratch slots might be used without an associated name? Or perhaps the key should be the slot number since that is required and unique?

Additionally I also assume this is just for documenting scratch variables intended to be read by other applications, not necessarily every scratch slot used?

Author: > Additionally I also assume this is just for documenting scratch variables intended to be read by other applications, not necessarily every scratch slot used?

Yeah exactly. The main purpose of having this field is so other contracts can easily access scratch variables. For example, TEALScript compiler uses scratch but this is only populated by user-defined scratch slots. I think forcing a name to be defined is fine because that's how it would be defined in the HLL. You can't have a property  that is just a number so a name needs to be defined. I definitely like adding a desc though

       * - literal: the value is a literal and should be passed directly as the argument
       * - method: The utf8 signature of the method in this contract to call to get the default value. If the method has arguments, they all must have default values. The method **MUST** be readonly so simulate can be used to get the default value.
       */
      source: "box" | "global" | "local" | "literal" | "method";
    };
  }>;
  /** Information about the method's return value */
  returns: {
    /** The type of the return value, or "void" to indicate no return value. The `struct` field should also be checked to determine if this return value is a struct. */
    type: ABIType;
    /** If the type is a struct, the name of the struct */
    struct?: StructName;
    /** Optional, user-friendly description for the return value */
    desc?: string;
  };
  /** an action is a combination of call/create and an OnComplete */
  actions: {
    /** OnCompletes this method allows when appID === 0 */
    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
    /** OnCompletes this method allows when appID !== 0 */
    call: (
      | "NoOp"
      | "OptIn"
      | "CloseOut"

>       | "ClearState"

Reviewer 4: Should `ClearState` be in here?

## Specification

### Contract Interface

Every application is described via the following interface which is an extension of the `Contract` interface described in [ARC-4](./arc-0004.md).

```ts
/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
interface Contract {
  /** The ARCs used and/or supported by this contract. All contracts implicitly support ARC4 and ARC56 */
  arcs: number[];
  /** A user-friendly name for the contract */
  name: string;
  /** Optional, user-friendly description for the interface */
  desc?: string;
  /**
   * Optional object listing the contract instances across different networks.
   * The key is the base64 genesis hash of the network, and the value contains
   * information about the deployed contract in the network indicated by the
   * key. A key containing the human-readable name of the network MAY be

>    * included, but the corresponding genesis hash key MUST also be define

Reviewer 5: ```suggestion
   * included, but the corresponding genesis hash key MUST also be defined
```

   */
  networks?: {
    [network: string]: {
      /** The app ID of the deployed contract in this network */
      appID: number;
    };
  };

>   /** Named structs use by the application. Each struct field appears in the same order as ABI encoding. */

Reviewer 5: ```suggestion
  /** Named structs used by the application. Each struct field appears in the same order as ABI encoding. */
```

      | "DeleteApplication"
    )[];
  };
  /** Information about the TEAL programs */
  sourceInfo?: {
    /** Approval program information */
    approval: ProgramSourceInfo;
    /** Clear program information */
    clear: ProgramSourceInfo;
  };
  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23 */
  source?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** The compiled bytecode for the application. MUST be omitted if included as part of ARC23 */
  byteCode?: {
    /** The approval program */
    approval: string;
    /** The clear program */
    clear: string;
  };
  /** Information used to get the given byteCode and/or PC values in sourceInfo. MUST be given if byteCode or PC values are present */
  compilerInfo?: {
    /** The name of the compiler */
    compiler: "algod" | "puya";
    /** Compiler version information */
    compilerVersion: {
      major: number;
      minor: number;
      patch: number;
      commitHash?: string;
    };
  };
  /** ARC-28 events that MAY be emitted by this contract */
  events?: Array<Event>;

>   /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) to their respective types and values (if applicable) */

Reviewer 5: ```suggestion
  /** A mapping of template variable names as they appear in the TEAL (not including TMPL_ prefix) to their respective types and values (if applicable) */
```

  templateVariables?: {
    [name: string]: {
      /** The type of the template variable */
      type: ABIType | AVMType | StructName;

>       /** If given, the the base64 encoded value used for the given app/program */

Reviewer 5: ```suggestion
      /** If given, the base64 encoded value used for the given app/program */
```

  desc?: string;
  /** The arguments of the method, in order */
  args: Array<{
    /** The type of the argument. The `struct` field should also be checked to determine if this arg is a struct. */
    type: ABIType;
    /** If the type is a struct, the name of the struct */
    struct?: StructName;
    /** Optional, user-friendly name for the argument */
    name?: string;
    /** Optional, user-friendly description for the argument */
    desc?: string;
    /** The default value that clients should use. */
    defaultValue?: {
      /** Base64 encoded bytes or uint64 */

>       data: string | bigint;

Reviewer 2: For JavaScript this will need to be `number` to not break deserialisation?

Author: I'm wondering if it's just easier to base64 encode uint64 rather than having the number? I could see this causing problems wiht JS otherwise.

Reviewer 2: Yeah I guess that's consistent? We'd have to specify how an `AVMUint64` would be encoded though (maybe we use ARC4 Uint64 encoding in that case so we can reuse `ABIType.from("uint64").decode(value)` rather than a custom decoding function?

Author: Yeah exactly I'm thinking we just use ARC4 encoding

Reviewer 2: The jsdoc should be explicit about that

Reviewer 4: It should also mention that `data` will be a UTF8 encoded ARC4 signature when source is `method`

Additionally what should `type` be in the case? Perhaps it should be optional and only for when source is  not `method`?

Author: Let me know how 20369ed9c82bc5aa00ae06985d4d24eb3f147861 looks

Reviewer 4: Looks good!

      /** How the data is encoded. This is the encoding for the data provided here, not the arg type */
      type: ABIType | AVMType;
      /** Where the default value is coming from
       * - box: The data key signifies the box key to read the value from
       * - global: The data key signifies the global state key to read the value from
       * - local: The data key signifies the local state key to read the value from (for the sender)

>        * - literal: the value is a literal and should be passed directly as the argument

Reviewer 2: As a suggestion, it could be the method name or method signature and the client is expected to use default values to populate any arguments. That implementation is simple, but powerful and I have an implementation ready to go (which is what it did for ARC-32).

>        * - literal: the value is a literal and should be passed directly as the argument

Author: Added in 3476dad

      */

>       source: 'box' | 'global' | 'local' | 'literal;

Reviewer 2: ```suggestion
    source: 'box' | 'global' | 'local' | 'literal';
```

    name?: string;
    /** Optional, user-friendly description for the argument */
    desc?: string;
    /** The default value that clients should use. */
    defaultValue?: {
      /** Base64 encoded bytes or uint64 */
      data: string | bigint;
      /** How the data is encoded. This is the encoding for the data provided here, not the arg type */

>       type: ABIType | AVMBytes;

Reviewer 2: This should probably have an `AVMUInt64` and is required to be that when the data is `bigint` and otherwise can't be that?

Author: Yeah I'm realizing we need to have AVMUint64 for state keys/values as well. Initially I was thinking it could always just be a native uint64 but there are cases where an app creator might reserve max bytes schema since you can store uint64 as bytes but not always the other way around. I will add AVMUint64 here and in state as well.

Author: I've added `AVMUint64` and `AVMType` as a uinion of all the AVM types and used `AVMType` where appropriate

type StructName = string;

/** Raw byteslice without the length prefixed that is specified in ARC-4 */
type AVMBytes = "AVMBytes";

/** A utf-8 string without the length prefix that is specified in ARC-4 */
type AVMString = "AVMString";

/** A 64-bit unsigned integer */
type AVMUint64 = "AVMUint64";

/** A native AVM type */
type AVMType = AVMBytes | AVMString | AVMUint64;

/** Information about a single field in a struct */

> interface StructField = {

Reviewer 2: ```suggestion
interface StructField {
```

type StructName = string;

/** Raw byteslice without the length prefixed that is specified in ARC-4 */
type AVMBytes = "AVMBytes";

> /** A string without the length prefix that is specified in ARC-4 */

Reviewer 2: Should we say it's a UTF-8 string?

/** Raw byteslice without the length prefixed that is specified in ARC-4 */
type AVMBytes = "bytes";

/** Information about a single field in a struct */
interface StructField = {
  /** The name of the struct field */
  name: string;
  /** The type of the struct field's value */

>   type: ABIType | StructName;

Reviewer 2: Should this be `ABIType | StructField[]` otherwise you have to create a unique name for all sub-values and makes the recursive logic harder?

Reviewer 4: Maybe `ABIType | StructName | StructField[]` as in some cases structs may be reused? And for languages without anonymous structured types (e.g. python) there should be a unique name already

Reviewer 2: Ah yeah, fair

### Template Variables

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: My main thinking was that it adds a lot of complexity to substitution. If we allow TMPL_ variables in `bytecblock` or `pushbytess` you have more complexity due to it being mixed in with other value, which may also contain `TMPL_` in non-template variable values.

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: I think the regex would still be pretty simple though?

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: I’m sure I’ve seen (PyTEAL?) contracts that have TMPL in constant block so it’s something that is in the wild…

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 4: I think if we don't prohibit `TMPL_` values appearing in constant blocks then that provides the opportunity for some other interesting solutions when debugging traces.

For example we can produce a debug map for a high level language that is independent of actual template values. By using pc offsets from the first op after the constant blocks, then when paired with a trace file we can determine the true pc values.

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 4: Regarding the regex, would it also need to include negative matches for other values inside the quote? e.g. don't match `"some strange TMPL_VALUE string"`

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: > I think if we don't prohibit TMPL_ values appearing in constant blocks then that provides the opportunity for some other interesting solutions when debugging traces.

Ah yeah that is interesting. That is why I emit `dissasembledTeal` in TEALScript, so you can use the dissasemlbed TEAL as the ultimate source of truth. 

> Regarding the regex, would it also need to include negative matches for other values inside the quote? e.g. don't match "some strange TMPL_VALUE string"

Right we'd also need to account for escaping (and escaping backslashes) which adds a lot of complexity

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: There's an existing regex that's been used in algokit utils for ages that's always seemed to work. It doesn't detect if the value is meant to be a uint vs bytes though (it does it based on whether it's passed a bigint/number or a uint8array.

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: The current implementations don't properly support bytecblock. If there was a string in the constant block with `TMPL_` then there would be a false match. I suppose we could simply forbid anything but hex in the constant block (in fact now that I think about it I'm not even sure what is supported)

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: Pretty sure the current algokit utils implementation supports bytecblock, that’s what all the unit tests work against funnily enough because pyteal uses it

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: But the question is not just if it supports bytecblock (because of course a regex would still work on a bytecblock), but whether or not it will erroneously replace `TMPL_` strings that are not intended to be a template variable. 

For example `bytecblock TMPL_FOO "TMPL_FOO"`

The first `TMPL_FOO` should be replaced but not the second. It is certainly an edge case though and maybe we think it's accetable?

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: And by acceptable I mean compilers don't let you use the string `TMPL_` anywhere. Otherwise the behavior would be unexpected.

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: I agree and I think it's a bit of an edge case.

Funnily enough, someone had a problem like that in python and the python utils has more complex code to deal with it (but it never made its way to typescript and noone complained).

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: Namely https://github.com/algorandfoundation/algokit-utils-py/blob/main/src/algokit_utils/deploy.py#L291

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 4: In terms of ARC-56, it should be sufficient to say what is allowed, but it doesn't need to describe how that is achieved (i.e. regex or otherwise) as that is an implementation detail?

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 4: Additionally I think we should still allow AVM uint64 template values. 

In cases where a fixed size uint value is required (i.e.  the cases\ described [here](https://github.com/algorandfoundation/ARCs/pull/258#issuecomment-2344642857)) then the template value can use one of the ARC4 numeric types to inform how a value should be encoded?

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: Wholeheartedly agree on both points.

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Author: iirc the Python code handles comments but not same-line strings. I could be wrong though. Either way it's a decent amount of complexity which is why I think it's important to outline a reference algorithim in the ARC. 

> Additionally I think we should still allow AVM uint64 template values.
>
> In cases where a fixed size uint value is required (i.e. the case described 
https://github.com/algorandfoundation/ARCs/pull/258#issuecomment-2344642857) then the template value can use one of the ARC4 numeric types to inform how a value should be encoded?

If we can agree on how source mapping should be done with them I agree, otherwise I don't think giving up human-readable errors is worth uint64 template variables. 

The options I see for source mapping are

A) Include decompliled TEAL lines in source map
B) Include bytecblock offset in source map

Both would require calls to algod awhich is not great and can be avoided (for integers at least) if we stick to byte-only, but it'd still be a problem for dynamic types. I think B is preferable because it doesn't depend on stability from the dissasembler

> Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.

Reviewer 2: Now we have a refined approach to template variables we should change the parts of this ARC that refer to needing to use pushbytes

### SourceInfo Interface

These interfaces give clients more information about the contract's source code.

```ts
interface ProgramSourceInfo {
  /** The source information for the program */
  sourceInfo: SourceInfo[];
  /** How the program counter offset is calculated
   * - none: The pc values in sourceInfo are not offset

>    * - cblocks: The pc values in sourceInfo are offset by PC of the last cblock at the top of the program

Reviewer 4: ```suggestion
   * - cblocks: The pc values in sourceInfo are offset by the PC of the first op following the last cblock at the top of the program
```

### Template Variables

Template variables are variables in the TEAL that should be substitued prior to compilation. The usage of the variable **MUST** appear in the TEAL starting with `TMPL_`. Template variables **MUST** be an argument to either `bytecblock` or `intcblock`. If a program has template variables, `bytecblock` and `intcblock` **MUST** be the first two opcodes in the program (unless one is not used).

#### Example

```js
#pragma version 10
bytecblock 0xdeadbeef TMPL_FOO
intcblock 0x12345678 TMPL_BAR
```

### Dynamic Template Variables

When a program has a template variable with a dynamic length, the `pcOffsetMethod` in `ProgramSourceInfo` **MUST** be `cblocks`. The `pc` value in each `SourceInfo` **MUST** be the pc determined at compilation minus the last `pc` value of the last `cblock` at compilation.

> When a client is leveraging a source map with `cblocks` as the `pcOffsetMethod`, it **MUST** determine the `pc` value by parsing the bytecode to get the PC value of the last `cblock` at the top of the program. See the reference implementation section for an example of how to do this.

Reviewer 4: ```suggestion
When a client is leveraging a source map with `cblocks` as the `pcOffsetMethod`, it **MUST** determine the `pc` value by parsing the bytecode to get the PC value of the first op following the last `cblock` at the top of the program. See the reference implementation section for an example of how to do this.
```
